processing/src/main/java/io/druid/query/groupby/epinephelinae/ConcurrentGrouper.java

```diff
@@ -292,19 +325,413 @@ private boolean isParallelSortAvailable()
+   * @return an iterator of the root of the combining tree
+   */
+  private Iterator<Entry<KeyType>> buildCombineTree(
+      List<Iterator<Entry<KeyType>>> sortedIterators,
+      Supplier<ByteBuffer> bufferSupplier,
+      AggregatorFactory[] combiningFactories,
+      int combineDegree,
+      List<Future> combineFutures,
+      List<String> dictionary
+  )
+  {
+    final int numIterators = sortedIterators.size();
+    if (numIterators > combineDegree) {
+      final int iteratorsPerChild = (numIterators + combineDegree - 1) / combineDegree; // ceiling
+      final List<Iterator<Entry<KeyType>>> childIterators = new ArrayList<>();
+      for (int i = 0; i < combineDegree; i++) {
+        childIterators.add(
+            buildCombineTree(
+                sortedIterators.subList(i * iteratorsPerChild, Math.min(numIterators, (i + 1) * iteratorsPerChild)),
+                bufferSupplier,
+                combiningFactories,
+                combineDegree,
+                combineFutures,
+                dictionary
+            )
+        );
+      }
+      return runCombiner(childIterators, bufferSupplier.get(), combiningFactories, combineFutures, dictionary).iterator();
+    } else {
+      return runCombiner(sortedIterators, bufferSupplier.get(), combiningFactories, combineFutures, dictionary).iterator();
+    }
+  }
+
+  private static ByteBuffer getSlice(ByteBuffer buffer, int sliceSize, int i)
+  {
+    final ByteBuffer slice = buffer.duplicate();
+    slice.position(sliceSize * i);
+    slice.limit(slice.position() + sliceSize);
+    return slice.slice();
+  }
+
+  private StreamingMergeSortedGrouper<KeyType> runCombiner(
+      List<Iterator<Entry<KeyType>>> iterators,
+      ByteBuffer combineBuffer,
+      AggregatorFactory[] combiningFactories,
+      List<Future> combineFutures,
+      List<String> dictionary
+  )
+  {
+    final Iterator<Entry<KeyType>> mergedIterator = Groupers.mergeIterators(iterators, keyObjComparator);
+    final SettableColumnSelectorFactory settableColumnSelectorFactory =
+        new SettableColumnSelectorFactory(aggregatorFactories);
+    final StreamingMergeSortedGrouper<KeyType> grouper = new StreamingMergeSortedGrouper<>(
+        Suppliers.ofInstance(combineBuffer),
+        combineKeySerdeFactory.factorizeWithDictionary(dictionary),
+        settableColumnSelectorFactory,
+        combiningFactories
+    );
+    grouper.init();
+
+    ListenableFuture future = grouperSorter.submit(() -> {
+      while (mergedIterator.hasNext()) {
+        final Entry<KeyType> next = mergedIterator.next();
+
+        settableColumnSelectorFactory.set(next.values);
+        grouper.aggregate(next.key); // grouper always returns ok or throws an exception
+        settableColumnSelectorFactory.set(null);
+      }
+
+      grouper.finish();
+    });
+
+    combineFutures.add(future);
+
+    return grouper;
+  }
+
   @Override
   public void close()
   {
     closed = true;
     for (Grouper<KeyType> grouper : groupers) {
```

Comment:
Maybe not close groupers if already closed = true

Reply:
Good point. Added.